### PR TITLE
[fix] torch.cat : cross-device check for out and input tensors

### LINKED ISSUE #52044: Illegal memory access when printing a CUDA tensor (problem with torch.stack(out=))
```python
import scipy.optimize
import torch

def linear_assignment(C):
    rowinds, colinds = zip(*list(map(scipy.optimize.linear_sum_assignment, C.cpu())))
    print(colinds[0].shape, colinds[0].dtype)
    # (8,) int64
    return torch.stack(list(map(torch.as_tensor, colinds)), out = torch.empty(C.shape[:2], dtype = torch.int64, device = C.device))
    # return torch.stack(list(map(torch.as_tensor, colinds))).to(C.device) # works


if __name__ == '__main__':
    C = torch.rand(31, 8, 8, device = 'cuda')
    P = linear_assignment(C)
    print(P.shape, P.dtype)
    # torch.Size([31, 8]) torch.int64
    print(P)
#    Traceback (most recent call last):
#  File "bug.py", line 13, in <module>
#    print(P)
#  File "/vadim/prefix/miniconda/lib/python3.8/site-packages/torch/tensor.py", line 179, in __repr__
#    return torch._tensor_str._str(self)
#  File "/vadim/prefix/miniconda/lib/python3.8/site-packages/torch/_tensor_str.py", line 372, in _str
#    return _str_intern(self)
#  File "/vadim/prefix/miniconda/lib/python3.8/site-packages/torch/_tensor_str.py", line 352, in _str_intern
#    tensor_str = _tensor_str(self, indent)
#  File "/vadim/prefix/miniconda/lib/python3.8/site-packages/torch/_tensor_str.py", line 241, in _tensor_str
#    formatter = _Formatter(get_summarized_data(self) if summarize else self)
#  File "/vadim/prefix/miniconda/lib/python3.8/site-packages/torch/_tensor_str.py", line 85, in __init__
#    value_str = '{}'.format(value)
#  File "/vadim/prefix/miniconda/lib/python3.8/site-packages/torch/tensor.py", line 534, in __format__
#    return self.item().__format__(format_spec)
#RuntimeError: CUDA error: an illegal memory access was encountered
```

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @ngimel

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/pytorch/pytorch/issues/52044 (`stack` dispatches to `cat`)

The way dispatcher works, currently this case happens only in CUDA kernel (CPU kernel is chosen if all inputs and out are on CPU). That is why the check is added only on the CUDA side.

### CODE REVIEW

File: aten/src/ATen/native/cuda/Shape.cu [hunk 1]
```diff
@@ -504,6 +504,11 @@ Tensor& cat_out_cuda(Tensor& out, TensorList inputs, int64_t dimension) {
                 t.device(), " and ", notSkippedTensor->device());
   }
 
+  TORCH_CHECK(
+      out.device() == notSkippedTensor->device(),
+      "All input tensors and out must be on the same device, but inputs are on ",
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/Shape.cu:509]: Remember to start the error message with the name of the op, so "torch.cat(): all input..."

File: test/test_tensor_creation_ops.py [hunk 2]
```diff
@@ -702,6 +702,10 @@ def test_cat_different_devices(self, devices):
                                     "input tensors must be on the same device"):
             torch.cat((cpu, cuda0))
 
+        with self.assertRaisesRegex(RuntimeError,
+                                    "All input tensors and out must be on the same device"):
+            torch.cat((cpu, cpu), out=cuda0)
```
Comment (mruberry) [inline test/test_tensor_creation_ops.py:707]: In this test it should be cat(cuda0, cuda0) with out=cuda1, and then a SEPARATE test that is @onlyCUDA should test cat(cpu, cpu) with out=cuda

**Review by mruberry — COMMENTED** [review]
Thanks for this fix, @kshitij12345, I made just a couple suggestions.

File: test/test_tensor_creation_ops.py (hunk 2 above)
Comment (mruberry) [reply test/test_tensor_creation_ops.py:707]: Also update that separate test with a call to stack directly to be 100% we're solving the issue. And - who knows - maybe in the future torch.stack() won't call torch.cat().

**Comment by kshitij12345** [comment]
@mruberry Gentle Ping :)

**Review by mruberry — APPROVED** [review]
Thanks @kshitij12345!

**Comment by mruberry** [comment]
A good follow-up would be making cat an OpInfo. We actually test for this in the generated out= tests now.

**Comment by kshitij12345** [comment]
> A good follow-up would be making cat an OpInfo. We actually test for this in the generated out= tests now.

Noted! Added to my list!

**Pull request closed without merging**